### PR TITLE
Manage Campaigns

### PR DESCRIPTION
**Pull request by Ap0c**
## Why are you doing this?

There are two possible types of AB test a user can be in on support-frontend:

- A **support-frontend test**, dictated by our AB test framework.
- A **dotcom test**, dictated by the dotcom AB test framework.

We have often wanted to modify the content of a page based upon the test to which a user has been allocated (this is usually the purpose of an AB test). For the support-frontend framework, we have a standardised way of doing this, using the `abtest.js` module and the redux store. However, for dotcom tests, changes have been applied ad-hoc (#116 and #197).

This PR standardises the process by introducing the concept of a 'campaign'. There are two reasons for this:

1. The way that we figure out whether a user is part of a dotcom AB test is by looking at the campaign code they have (`INTCMP` query param) when they arrive on our site.
2. It could be argued that a user's campaign is a superset of their dotcom AB test allocation as far as we're concerned. In simpler terms, we're currently looking at modifying the page content based upon the dotcom AB test, but we may want to modify a user's experience more generally by campaign in the future (contributions have [already done this](https://github.com/guardian/contributions-frontend/pull/264)).

[**Trello Card**](https://trello.com/c/O7KgTKLW/867-refactor-create-a-test-helper-in-support-frontend-where-we-can-centralise-all-the-tests-and-put-the-belongtotest-function)

## Changes

- Created a new `tracking` subdirectory in shared helpers, and moved `ga.js` into it.
- Pulled the `intCmp` and `refpvid` modules into a new `guTracking` module.
- Created a new `campaigns` object in the `guTracking` module, and provided methods for retrieving information about a user's campaign.
- Updated `subscriptionsLinks` to build subs links based upon the user's campaign.

### CODE REVIEW

File: assets/helpers/tracking/guTracking.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// @flow
+
+// ----- Campaigns ----- //
+
+const campaigns : {
+  [string]: string[],
+} = {
+  baseline_test: [
+    'gdnwb_copts_memco_sandc_support_baseline_support_banner',
```
Comment (rupertbates) [inline assets/helpers/tracking/guTracking.js]: Maybe extract 'gdnwb_copts_memco_sandc_support_baseline_support_' out into a const and then append the final part of the code onto it, for readability and ease of maintenance

File: assets/helpers/tracking/guTracking.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  return campaigns[campaign] && campaigns[campaign].includes(intCmp);
+
+}
+
+// Retrieves the user's campaign, if known, from the intCmp.
+function getCampaign(intCmp: string): ?Campaign {
+
+  let campaignName = null;
+
+  Object.keys(campaigns).forEach((campaign) => {
```
Comment (rupertbates) [inline assets/helpers/tracking/guTracking.js]: You should be able to use .filter() here instead of forEach and then you don't need the campaignName variable.

File: assets/helpers/tracking/guTracking.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
+// @flow
+
+// ----- Campaigns ----- //
+
+const campaigns : {
+  [string]: string[],
+} = {
+  baseline_test: [
```
Comment (davidfurey) [inline assets/helpers/tracking/guTracking.js]: This seems a little bit confusing.  intCmp is internal campaign code, which implies one campaign per code, but here you have `campaign` as a collection of campaign codes.  If this concept is used elsewhere in the business, that's fine.  Otherwise I'd prefer we used a different name.  Is the collective noun for campaigns a conflict?

Or since this seems to map to tests should we call it `const tests` that maps from test to campaign code?

File: assets/helpers/tracking/guTracking.js (hunk 2 above)
Comment (Ap0c) [reply assets/helpers/tracking/guTracking.js]: Nice, updated.

File: assets/helpers/tracking/guTracking.js (hunk 1 above)
Comment (Ap0c) [reply assets/helpers/tracking/guTracking.js]: Do you mean something like this?

```js
const prefix = 'gdnwb_copts_memco_sandc_support_baseline_support_';

const campaigns : {
  [string]: string[],
} = {
  baseline_test: [
    'banner',
    'epic',
    'liveblog',
    'header_become_supporter',
    'header_subscribe',
    'side_menu_become_supporter',
    'side_menu_subscribe',
  ].map(suffix => `${prefix}${suffix}`)
};
```

File: assets/helpers/tracking/guTracking.js (hunk 3 above)
Comment (Ap0c) [reply assets/helpers/tracking/guTracking.js]: Here I'm seeing it as our 'baseline test' is a type of campaign that we're running, since we're using the internal campaign codes to track it.

I'm reluctant to use the word 'test' because:

a) I feel like it's a bit generic and is starting to get overloaded.
b) I can see the application as being wider than just tests, we are probably going to want to track and modify content based upon campaigns generally, as contributions do.

File: assets/helpers/tracking/guTracking.js (hunk 1 above)
Comment (rupertbates) [reply assets/helpers/tracking/guTracking.js]: Yes, nice.

Comment (Ap0c) [reply assets/helpers/tracking/guTracking.js]: Updated to use @JustinPinner's solution, by popular vote, as discussed online elsewhere (as opposed to offline).

**Comment by JustinPinner** [comment]
Seems sensible to me 👍

File: assets/helpers/tracking/guTracking.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+}
+
+
+// ----- Reducers ----- //
+
+// Since nothing changes the INTCMP, this reducer does not handle any actions.
+const intCmpReducer = (state: ?string = null): ?string => state;
+
+const refpvidReducer = (state: ?string = null): ?string => state;
```
Comment (svillafe) [inline assets/helpers/tracking/guTracking.js]: Is it possible to combine these two reducers in one reducer called `guTrackingReducer` which state contains two fields? In that way, we would import only one reducer in other files and in that way have less code around.

File: assets/helpers/tracking/guTracking.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  [string]: string[],
+} = {
+  baseline_test: [
+    `${baselineTestPrefix}_banner`,
+    `${baselineTestPrefix}_epic`,
+    `${baselineTestPrefix}_liveblog`,
+    `${baselineTestPrefix}_header_become_supporter`,
+    `${baselineTestPrefix}_header_subscribe`,
+    `${baselineTestPrefix}_side_menu_become_supporter`,
+    `${baselineTestPrefix}_side_menu_subscribe`,
```
Comment (svillafe) [inline assets/helpers/tracking/guTracking.js]: In the future, it would be nice to standardise this channels and create a type which validates them.

File: assets/pages/bundles-landing/helpers/subscriptionsLinks.js [hunk 6]
```diff
@@ -23,29 +29,20 @@ const defaultPromos: PromoCodes = {
...
-      paper: 'p/NJ6CCQLD4',
-      paperDig: 'p/NJ6CCSVMC',
-    },
+const customPromos : {
+  [Campaign]: PromoCodes,
+} = {
+  baseline_test: {
+    digital: 'p/DJ6CCJZLL',
+    paper: 'p/NJ6CCQLD4',
+    paperDig: 'p/NJ6CCSVMC',
```
Comment (svillafe) [inline assets/pages/bundles-landing/helpers/subscriptionsLinks.js]: In order to have everything which is related to campaign codes in one place, we can probably move this inside the `guTracking` folder.
